kratos/utilities/strategies_utilities.h

```diff
@@ -0,0 +1,198 @@
+
+        double* Avalues = rA.value_data().begin();
+        std::size_t* Arow_indices = rA.index1_data().begin();
+
+        // The diagonal considered
+        const double scale_factor = GetScaleNorm<TSparseSpace>(rModelPart, rA, ScalingDiagonal);
+
+        // Detect if there is a line of all zeros and set the diagonal to a 1 if this happens
+        IndexPartition<std::size_t>(system_size).for_each([&](std::size_t Index){
+            bool empty = true;
+
+            const std::size_t col_begin = Arow_indices[Index];
+            const std::size_t col_end = Arow_indices[Index + 1];
+
+            for (std::size_t j = col_begin; j < col_end; ++j) {
+                if(Avalues[j] != 0.0) {
+                    empty = false;
+                    break;
+                }
+            }
+
+            if(empty) {
+                rA(Index, Index) = scale_factor;
+                rb[Index] = 0.0;
+            }
+        });
+
+        return scale_factor;
+    }
+
+    /**
+     * @brief This method returns the scale norm considering for scaling the diagonal
+     * @param rModelPart The problem model part
+     * @param rA The LHS matrix
+     * @param ScalingDiagonal The type of caling diagonal considered
+     * @return The scale norm
+     */
+    template<class TSparseSpace>
+    static double GetScaleNorm(
+        ModelPart& rModelPart,
+        typename TSparseSpace::MatrixType& rA,
+        const SCALING_DIAGONAL ScalingDiagonal = SCALING_DIAGONAL::NO_SCALING
+        )
+    {
+        switch (ScalingDiagonal) {
+            case SCALING_DIAGONAL::NO_SCALING:
+                return 1.0;
+            case SCALING_DIAGONAL::CONSIDER_PRESCRIBED_DIAGONAL: {
+                const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+                KRATOS_ERROR_IF_NOT(r_current_process_info.Has(BUILD_SCALE_FACTOR)) << "Scale factor not defined at process info" << std::endl;
+                return r_current_process_info.GetValue(BUILD_SCALE_FACTOR);
+            }
+            case SCALING_DIAGONAL::CONSIDER_NORM_DIAGONAL:
+                return GetDiagonalNorm<TSparseSpace>(rA)/static_cast<double>(rA.size1());
+            case SCALING_DIAGONAL::CONSIDER_MAX_DIAGONAL:
+                return GetMaxDiagonal<TSparseSpace>(rA);
+//                 return TSparseSpace::TwoNorm(rA)/static_cast<double>(rA.size1());
+            default:
+                return GetMaxDiagonal<TSparseSpace>(rA);
+        }
+    }
+
+    /**
+     * @brief This method returns the diagonal norm considering for scaling the diagonal
+     * @param rA The LHS matrix
+     * @return The diagonal norm
+     */
+    template<class TSparseSpace>
+    static double GetDiagonalNorm(typename TSparseSpace::MatrixType& rA)
+    {
+        double diagonal_norm = 0.0;
+        diagonal_norm = IndexPartition<std::size_t>(TSparseSpace::Size1(rA)).for_each<SumReduction<double>>([&](std::size_t Index){
+            return std::pow(rA(Index,Index), 2);
```

Comment:
accessing the `compressed_matrix` with `()` is slow and I am not sure what happens if the entry does not exist (should hopefully not be the case but better safe than sorry)

This is how it should be accessed: https://github.com/KratosMultiphysics/Kratos/blob/c7f5682de5b3c06819005e332d563923f70b27c3/kratos/spaces/ublas_space.h#L331-L334

=> check also the corresponding test I wrote

There are some resources available on google, I think it is worth doing it properly here

@RiccardoRossi maybe you have more suggestions, I think you know this well

Reply:
This code is just ported from the BlockBuilderAndSolver: https://github.com/KratosMultiphysics/Kratos/blob/c7f5682de5b3c06819005e332d563923f70b27c3/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h#L1748 and followings

I can take a look to make it faster, but then it will lose generality (it will work only for a specific sparse space)

Reply:
BTW, UblasSpace has not being refactored with ParallelUtilities yet

Reply:
Optimized!